src/jws/flattened.rs

```diff
@@ -0,0 +1,542 @@
+//! Flattened JWS signatures: see RFC 7515 section 7.2.2
+//! Flattened signatures are JSON (unlike compact signatures),
+//! and support a single signature protecting a set of headers and a
+//! payload.
+//! The RFC specifies unprotected headers as well, but this implementation
+//! doesn't support them.
+
+use super::{Header, RegisteredHeader, Secret};
+use crate::errors::{Error, ValidationError};
+use crate::jwa::SignatureAlgorithm;
+use crate::serde_custom;
+use data_encoding::BASE64URL_NOPAD;
+use serde::{de::DeserializeOwned, Deserialize, Deserializer, Serialize};
+
+// Not using CompactPart::to_bytes here, bounds are overly restrictive
+fn serialize_header<H: Serialize>(header: &Header<H>) -> Vec<u8> {
+    serde_json::to_vec(header).unwrap()
+}
+
+/// Warning: pay attention to parameter order
+// Note: this is valid UTF-8, but gets used as bytes later
+fn signing_input(protected_header: &[u8], payload: &[u8]) -> Vec<u8> {
+    let hlen = BASE64URL_NOPAD.encode_len(protected_header.len());
+    let plen = BASE64URL_NOPAD.encode_len(payload.len());
+    let mut r = Vec::with_capacity(hlen + plen + 1);
+    r.append(&mut BASE64URL_NOPAD.encode(protected_header).into_bytes());
+    r.push(b'.');
+    r.append(&mut BASE64URL_NOPAD.encode(payload).into_bytes());
+    r
+}
+
+/// Data that can be turned into a JWS
```

Comment:
Please provide some examples on how this can be used. Also, please indicate that this is for flattened JWS.